### ISSUE TITLE
Attribute defaults

### ISSUE THREAD

**Issue by dschult** [pygraphviz/pygraphviz#303]
Attributes in pygraphviz are slightly different from those in Graphviz and can be confusing. I believe the confusion stems from Graphviz's handling of default values, but pygraphviz's reporting of attributes doesn't help. This issue is an attempt to describe how attributes work. In the end I don't think we should change anything. So I will likely close this issue immediately but leave it for people to find when searching.

Graphviz yields more attributes for nodes/edges than pygraphviz does with e.g. `A.get_node(3).attr.iteritems()`. 
Pygraphviz hides any attributes that have the same value as the defaults.

In Graphviz, the default attribute values are only used to set attribute values for **new** nodes/edges.
The values of existing attributes are not changed when defaults are set to a new value.
In pygraphviz it looks like the attributes are created or destroyed when you change the default values because attributes with values equal to the default value are not reported. So if node 3 has color "red", it will not report a color attribute when the default node attribute for color is "red", but it will suddenly report a color attribute with value "red" when the default value is set to "blue".

My intuition for default values is that so long as a specific value has never been set for a node, that node should take on the current default value for an attribute.  That is, you change the default to "red" and all nodes show as "red", then changing the default to "blue" should make all nodes "blue" until the default is reset or a specific color is set for that node. I guess the idea that setting a default value only changes newly created nodes is uncomfortable for me -- certainly different from matplotlib defaults.

Questions: 
1) Should pygraphviz store default values separately from Graphviz so unset nodes can take on the default values w/o being reset?
2) Should pygraphviz report all attributes whether they have the same value as the default or not (as graphviz does)?

My Current Answers:
1) I think this moves pygraphviz away from being a wrapper of graphviz. I am reluctant to do that without a bigger plan for what it would become.
2) Making pygraphviz report more directly what graphviz reports aligns with being a wrapper of the library. It doesn't actually affect anything outside of the `iteritems` method because that method is never used within the library except for testing. This would not be backward compatible though, would change any tests that use attribute iteritems and could be difficult to implement. So I'm +1 for leaving it as it is...

tl;dr

### Graphviz
**New attributes**:
In Graphviz, creating a new attribute for any node creates that attribute for all nodes.
All other nodes then have that attribute with an empty string "" as the value.
A new default attribute is also created with empty string for a value.

Similarly, creating a default value for a new node attribute creates that attribute for all nodes.
Every node then reports that attribute with value equal to the default attribute value.

**Existing Attributes**:
In Graphviz, changing the value of a node's attribute just changes that node's attribute value as expected.
Changing the value of an existing default attribute does not change any existing node/edge attribute values.
It only changes the default for nodes added later.  This difference between new and existing could cause confusion e.g.:
```python
A=pgv.AGraph()
A.add_path([1,2,3])   # drawing now creates black nodes
A.node_attr['color']="red"  # drawing now creates red nodes
A.node_attr['color']="blue"  # drawing now still creates red nodes
A.add_node(4)  # node "4" will be blue but nodes 1, 2, 3 will be red.
```
Creating the attribute by setting it on a node before the default value is set gives different results.
```python
A=pgv.AGraph()
A.add_path([1,2,3])
A.get_node(3).attr["color"] = "blue"   # sets the color attr for all other nodes and default to ""
A.node_attr['color']="red"  # Changes default for new nodes -- not existing nodes. 
#  Drawing now creates black for nodes 1 and 2 and blue for node 3.
```

###Pygraphviz
pygraphviz does not report any attributes when the attribute value matches the default value.
```python
>>> A=pgv.AGraph()
>>> A.add_path([1,2,3])
>>> A.node_attr['color']="red"  # drawing now creates red nodes
>>> list(A.get_node(3).attr.iteritems())  # No attributes appear to be set
[]
>>> A.node_attr['color']="blue"  # drawing now still creates red nodes
>>> list(A.get_node(3).attr.iteritems())  # Now attributes do appear for each node. But the nodes haven't changed.
[('color', 'red')]
>>> A.add_node(4)  # node "4" will be blue but nodes 1, 2, 3 will be red.
>>> list(A.get_node(3).attr.iteritems())  #  
[('color', 'red')]
>>> list(A.get_node(4).attr.iteritems())  # 
[('color', 'blue')]
```

**Comment by dschult** [comment]
I'm going to close this now as it ends up proposing no changes. But the description of how attributes work might be helpful to be in the searchable issues.
